File: CodeBattleNet/CodeBattleNetLibrary/GameClientBomberman.cs
using System;
using WebSocket4Net;

namespace CodeBattleNetLibrary
{
	public class GameClientBomberman
	{
		private readonly WebSocket _socket;
		private event Action OnUpdate;

		public BombermanBlocks[,] Map { get; private set; }
		public int MapSize { get; private set; }
		public int PlayerX { get; private set; }
		public int PlayerY { get; private set; }

		public GameClientBomberman(string server, string userEmail, string code)
		{
			MapSize = 0;

			_socket =
				new WebSocket(
					$"ws://{server}/codenjoy-contest/ws?user={userEmail}&code={code}");
            _socket.Opened += new EventHandler(onOpened);
            _socket.Error += new EventHandler<SuperSocket.ClientEngine.ErrorEventArgs>(onError);
            _socket.Closed += new EventHandler(onClose);
            _socket.MessageReceived += (s, e) => { ParseField(e.Message); };
           
		}

		public void Run(Action handler)
		{
			OnUpdate += handler;
			_socket.Open();
		}

		public void Up(BombAction bombAction = BombAction.None)
		{
			send(
				$"{(bombAction == BombAction.BeforeTurn ? "ACT," : "")}UP{(bombAction == BombAction.AfterTurn ? ",ACT" : "")}"
				);
		}

		public void Down(BombAction bombAction = BombAction.None)
		{
            send(
				$"{(bombAction == BombAction.BeforeTurn ? "ACT," : "")}DOWN{(bombAction == BombAction.AfterTurn ? ",ACT" : "")}"
				);
		}

		public void Right(BombAction bombAction = BombAction.None)
		{
            send(
				$"{(bombAction == BombAction.BeforeTurn ? "ACT," : "")}RIGHT{(bombAction == BombAction.AfterTurn ? ",ACT" : "")}"
				);
		}

		public void Left(BombAction bombAction = BombAction.None)
		{
            send(
				$"{(bombAction == BombAction.BeforeTurn ? "ACT," : "")}LEFT{(bombAction == BombAction.AfterTurn ? ",ACT" : "")}"
				);
		}

		public void Act()
		{
            send("ACT");
		}

		public void Blank()
		{
            send("");
		}

		private void ParseField(string rawField)
		{
			rawField = rawField.Substring(6);
			int size = (int) Math.Sqrt(rawField.Length);
			if (MapSize != size)
			{
				Map = new BombermanBlocks[size, size];
				MapSize = size;
			}

			int rawPosition = 0;
			for (int j = 0; j < size; j++)
			{
				for (int i = 0; i < size; i++)
				{
					Map[i, j] = CharToBlock(rawField[rawPosition]);

					if (IsPlayerCoords(Map[i, j]))
					{
						PlayerX = i;
						PlayerY = j;
					}

					rawPosition++;
				}
			}

			OnUpdate?.Invoke();
		}

        private void onOpened(object sender, EventArgs e)
        {
            Console.WriteLine("Connection established");
        }

        private void onError(object sender, EventArgs e)
        {
            Console.WriteLine("### error ###\n" + e.ToString());
        }

        private void onClose(object sender, EventArgs e)
        {
            Console.WriteLine("### disconnected ###");
        }

        private void send(string msg)
        {
            Console.WriteLine($"Sending: {msg}");
            _socket.Send(msg);
        }

		protected bool IsPlayerCoords(BombermanBlocks block) => block == BombermanBlocks.Bomberman ||
		                                                        block == BombermanBlocks.BombBomberman ||
		                                                        block == BombermanBlocks.DeadBomberman;

		protected BombermanBlocks CharToBlock(char c) =>
			Enum.IsDefined(typeof (BombermanBlocks), (int) c)
				? (BombermanBlocks) c
				: BombermanBlocks.Unknown;
	}
}

File: CodeBattleCpp/CodeBattleCppLibrary/GameClientBomberman.cpp
#include "GameClientBomberman.h"

#include <iostream>

GameClientBomberman::GameClientBomberman(std::string _server, std::string _userEmail, std::string _code)
{
	map = nullptr;
	map_size = 0;

	path = "ws://" + _server + "/codenjoy-contest/ws?user=" + _userEmail + "&code=" + _code;

	is_running = false;
}

GameClientBomberman::~GameClientBomberman()
{
	is_running = false;
	work_thread->join();
}

void GameClientBomberman::Run(std::function<void()> _message_handler)
{
	is_running = true;
	work_thread = new std::thread(&GameClientBomberman::update_func, this, _message_handler);
}

void GameClientBomberman::update_func(std::function<void()> _message_handler)
{
#ifdef _WIN32
	WSADATA wsaData;

	if (WSAStartup(MAKEWORD(2, 2), &wsaData))
		throw new std::exception("WSAStartup Failed.\n");
	else
		std::cout << "Connection established" << std::endl;
#endif

	web_socket = easywsclient::WebSocket::from_url(path);
	if (web_socket == nullptr)is_running = false;
	while (is_running)
	{
		web_socket->poll();
		web_socket->dispatch([&](const std::string &message)
		{
#ifdef _WIN32
			int size_needed = MultiByteToWideChar(CP_UTF8, 0, &message[0], (int)message.size(), NULL, 0);
			std::wstring wmessage(size_needed, 0);
			MultiByteToWideChar(CP_UTF8, 0, &message[0], (int)message.size(), &wmessage[0], size_needed);
#else
			std::wstring wmessage(message.begin(), message.end());
#endif
			
			uint32_t size = sqrt(wmessage.size() - 6);
			if (map_size != size)
			{
				if (map_size != 0)
				{
					for (uint32_t j = 0; j < map_size; j++)
						delete[] map[j];
					delete[] map;
				}
				map_size = size;

				map = new BombermanBlocks*[map_size];
				for (uint32_t j = 0; j < map_size; j++)
				{
					map[j] = new BombermanBlocks[map_size];
					for (uint32_t i = 0; i < map_size; i++)
					{
						map[j][i] = BombermanBlocks::Unknown;
					}
				}
			}

			uint32_t chr = 6;
			for (uint32_t j = 0; j < map_size; j++)
			{
				for (uint32_t i = 0; i < map_size; i++)
				{
					map[j][i] = (BombermanBlocks)wmessage[chr];
					chr++;

					if (map[j][i] == BombermanBlocks::Bomberman || map[j][i] == BombermanBlocks::BombBomberman || map[j][i] == BombermanBlocks::DeadBomberman)
					{
						player_x = i;
						player_y = j;
					}
				}
			}

			_message_handler();
		});
	}
	if (web_socket)web_socket->close();

#ifdef _WIN32
	WSACleanup();
#endif
}
